File: tests/xcalablemp/NOT-IMPLEMENTED/39/in11.f90
! testp115.f
! barrier指示文のテスト

      program main
      include 'xmp_lib.h'
      integer,parameter:: N=1000
!$xmp nodes p(*)
!$xmp template t(N)
!$xmp distribute t(block) onto p
      integer a(N), procs, w
!$xmp align a(i) with t(i)
      character(len=3) result

      procs = xmp_num_nodes()
      if(mod(N,procs) .eq. 0) then
         w = N/procs
      else
         w = N/procs+1
      endif
      
      do j=2, procs
         if(xmp_node_num() .eq. 1) then
            do i=1, w
               a(i) = (j-1)*w+i
            enddo
         endif
!$xmp barrier on p(:)
         if(xmp_node_num() .eq. j) then
!$xmp gmove in
            a((j-1)*w+1:j*w) = a(1:w)
         endif
      enddo

      if(xmp_node_num() .eq. 1) then
         do i=1, w
            a(i) = i
         enddo
      endif
      
      result = 'OK '
!$xmp loop on t(i)
      do i=1, N
         if(a(i) .ne. i) result = 'NG '
      enddo

      print *, xmp_node_num(), 'testp115.f ', result

      end

File: tests/XMP/NOT-IMPLEMENTED/39/in11.c
/*testp115.c*/
/*barrier指示文のテスト*/
#include<mpi.h>
#include<xmp.h>
#include<stdio.h>
#include<stdlib.h>   
#include<unistd.h>
static const int N=1000;
#pragma xmp nodes p(*) 
#pragma xmp template t(0:N-1)
#pragma xmp distribute t(block) onto p
int a[N],procs,w;
int i,j;
#pragma xmp align a[i] with t(i)
char *result;
int main(void){

   procs = xmp_num_nodes();
    
   for(j=1;j<procs;j++){
      if((N%procs)==0){
         w = N/procs;
      }else{
         w = N/procs+1;
         w = (w < (N-w*(j-1)))? w: N-w*(j-1);
      }
      if(xmp_node_num() == 1){
         for(i=0;i<w;i++){
            a[i] = j*w+i;
         }
      }
   
#pragma xmp barrier on p(:)
#pragma xmp task on p(j+1)
      {
#pragma xmp gmove in 
         a[j*w:w] = a[0:w];
      }
   }
   if(xmp_node_num() == 1){
      for(i=0;i<w;i++){
         a[i] = i+1;
      }
   }
   result = "OK";
#pragma xmp loop on t(i)
   for(i=0;i<N;i++){
      if(a[i] != i) result = "NG";
   }

   printf("%d %s %s\n",xmp_node_num(),"testp115.c",result);
   return 0;
}
      
         
     
   

